cpp/src/arrow/io/file.cc

```diff
@@ -152,20 +152,39 @@ static inline int64_t lseek64_compat(int fd, int64_t pos, int whence) {
 #endif
 }
 
+#if defined(_MSC_VER)
+static inline Status Utf8ToUtf16(const std::string& input, std::wstring &result) {
```

Comment:
Note that the Google style guide forbids mutable reference arguments: https://google.github.io/styleguide/cppguide.html#Reference_Arguments, so this would need to be `std::wstring* result`